File: gunhoflash/container-with-most-water.ts
function maxArea(heights: number[]): number {
  let answer = 0;
  let leftIndex = 0;
  let rightIndex = heights.length - 1;

  while (leftIndex < rightIndex) {
    const leftHeight = heights[leftIndex];
    const rightHeight = heights[rightIndex];

    const area = (rightIndex - leftIndex) * Math.min(leftHeight, rightHeight);
    answer = Math.max(answer, area);

    if (rightHeight > leftHeight) {
      leftIndex++;
    } else {
      rightIndex--;
    }
  }

  return answer;
}

File: gunhoflash/container-with-most-water.py
from typing import List

class Solution:
  def maxArea(self, heights: List[int]) -> int:
    answer = 0
    leftIndex = 0
    rightIndex = len(heights) - 1

    while leftIndex < rightIndex:
      leftHeight = heights[leftIndex]
      rightHeight = heights[rightIndex]

      area = (rightIndex - leftIndex) * min(leftHeight, rightHeight)
      answer = max(answer, area)

      if rightHeight > leftHeight:
        leftIndex += 1
      else:
        rightIndex -= 1

    return answer
